pkg/k8s/client/object_tracker.go

```diff
@@ -0,0 +1,648 @@
+	wtxn.Commit()
+	if err != nil {
+		// Impossible
+		panic(err)
+	}
+
+	var fieldSelector fields.Selector
+	version := uint64(0)
+	if len(opts) > 0 && opts[0].ResourceVersion != "" {
+		opt := opts[0]
+		version, err = strconv.ParseUint(opt.ResourceVersion, 10, 64)
+		if err != nil {
+			return nil, err
+		}
+
+		if opt.FieldSelector != "" {
+			fieldSelector, err = fields.ParseSelector(opt.FieldSelector)
+			if err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	s.log.Debug("Watch",
+		logfieldClientset, s.domain,
+		logfieldGVR, gvr,
+		logfields.K8sNamespace, ns,
+		logfieldResourceVersion, version)
+
+	w := &statedbWatch{
+		clientset:     s.domain,
+		log:           s.log,
+		version:       version,
+		gvr:           gvr,
+		ns:            ns,
+		db:            s.db,
+		iter:          changeIter,
+		stop:          make(chan struct{}),
+		stopped:       make(chan struct{}),
+		events:        make(chan watch.Event, 1),
+		fieldSelector: fieldSelector,
+	}
+	go w.feed()
+
+	return w, nil
+}
+
+type statedbWatch struct {
+	clientset     string
+	log           *slog.Logger
+	gvr           schema.GroupVersionResource
+	ns            string
+	version       statedb.Revision
+	db            *statedb.DB
+	iter          statedb.ChangeIterator[object]
+	stop          chan struct{}
+	stopped       chan struct{}
+	events        chan watch.Event
+	fieldSelector fields.Selector
+}
+
+// ResultChan implements watch.Interface.
+func (w *statedbWatch) ResultChan() <-chan watch.Event {
+	return w.events
+}
+
+func (w *statedbWatch) feed() {
+	defer close(w.stopped)
+	defer close(w.events)
+	seen := sets.New[string]()
+	for {
+		changes, changesWatch := w.iter.Next(w.db.ReadTxn())
+		for change := range changes {
+			if change.Revision <= w.version {
+				continue
+			}
+			obj := change.Object
+			if obj.domain != w.clientset {
+				continue
+			}
+			if (w.ns != "" && obj.Namespace != w.ns) || obj.gvr != w.gvr {
+				continue
+			}
+
+			if w.fieldSelector != nil {
+				if !objectMatchesFieldSelector(obj.o, w.fieldSelector) {
+					w.log.Debug("Watch: Skipping event due to FieldSelector mismatch",
+						logfieldFieldSelector, w.fieldSelector)
+					continue
+				}
+			}
+
+			var ev watch.Event
+			ev.Object = obj.o.DeepCopyObject()
+
+			switch {
+			case change.Deleted:
+				ev.Type = watch.Deleted
+				seen.Delete(obj.Name)
+			case seen.Has(obj.Name):
+				ev.Type = watch.Modified
+			default:
+				ev.Type = watch.Added
+				seen.Insert(obj.Name)
+			}
+			w.log.Debug(
+				"Event",
+				logfieldGVR, obj.gvr,
+				logfields.K8sNamespace, obj.Namespace,
+				logfields.Name, obj.Name,
+				logfields.Type, ev.Type,
+				logfieldResourceVersion, change.Revision,
+				logfields.Object, ev.Object,
+			)
+			select {
+			case w.events <- ev:
+			case <-w.stop:
+				return
+			}
+		}
+		select {
+		case <-w.stop:
+			return
+		case <-changesWatch:
+		}
+
+	}
+}
+
+// Stop implements watch.Interface.
+func (w *statedbWatch) Stop() {
+	close(w.stop)
```

Comment:
Nit: Looking through other implementations of this, usually Stop() is implemented to be idempotent to not crash if called twice (however the interface doesn't specify this so either way is probably fine).